src/wasi_unstable/mod.rs

```diff
@@ -526,99 +513,231 @@ pub fn path_symlink(old_path: &[u8], fd: Fd, new_path: &[u8]) -> Errno {
+#[inline]
+pub fn fd_prestat_dir_name(fd: Fd, path: &mut [u8]) -> Result<(), Error> {
+    unsafe_wrap0!{ __wasi_fd_prestat_dir_name(fd, path.as_mut_ptr(), path.len()) }
+}
+
+#[cfg(feature = "alloc")]
+use alloc::{vec::Vec, string::String};
+
+#[cfg(feature = "alloc")]
+pub fn get_args() -> Result<Vec<Vec<u8>>, Error> {
+    use alloc::vec;
+
+    let (mut argc, mut argv) = (0, 0);
+    let ret = unsafe { __wasi_args_sizes_get (&mut argc, &mut argv) };
+    if let Some(err) = NonZeroU16::new(ret) { return Err(err); }
+
+    let mut argc = vec![core::ptr::null_mut::<u8>(); argc];
+    let mut argv = vec![0u8; argv];
+    let ret = unsafe { __wasi_args_get(argc.as_mut_ptr(), argv.as_mut_ptr()) };
+    if let Some(err) = NonZeroU16::new(ret) { return Err(err); }
+
+    fn cstr2vec(ptr: *const u8) -> Vec<u8> {
+        let mut n: usize = 0;
+        loop {
+            unsafe {
+                // replace with memchr?
+                if *ptr.offset(n as isize) == 0 {
+                    return core::slice::from_raw_parts(ptr, n).to_vec();
+                }
+            }
+            n += 1;
+        }
+    }
+
+    Ok(argc.into_iter().map(|p| cstr2vec(p as *const u8)).collect())
 }
 
-// TODO: Safe interfaces to the args and environ functions
-/*
-pub fn args_get(argv: *mut *mut u8, argv_buf: *mut u8) -> Errno {}
-pub fn args_sizes_get(argc: *mut usize, argv_buf_size: *mut usize) -> Errno {}
-pub fn environ_get(environ: *mut *mut u8, environ_buf: *mut u8) -> Errno {}
-pub fn environ_sizes_get(environ_count: *mut usize, environ_buf_size: *mut usize) -> Errno {}
-*/
+#[cfg(feature = "alloc")]
+pub fn get_environ() -> Result<Vec<Vec<u8>>, Error> {
+    use alloc::vec;
+
+    let (mut argc, mut argv) = (0, 0);
+    let ret = unsafe { __wasi_environ_sizes_get (&mut argc, &mut argv) };
+    if let Some(err) = NonZeroU16::new(ret) { return Err(err); }
+
+    let mut argc = vec![core::ptr::null_mut::<u8>(); argc];
+    let mut argv = vec![0u8; argv];
+    let ret = unsafe { __wasi_environ_get(argc.as_mut_ptr(), argv.as_mut_ptr()) };
+    if let Some(err) = NonZeroU16::new(ret) { return Err(err); }
+
+    fn cstr2vec(ptr: *const u8) -> Vec<u8> {
+        let mut n: usize = 0;
+        loop {
+            unsafe {
+                if *ptr.offset(n as isize) == 0 {
+                    return core::slice::from_raw_parts(ptr, n).to_vec();
+                }
+            }
+            n += 1;
+        }
+    }
+
+    Ok(argc.into_iter().map(|p| cstr2vec(p as *const u8)).collect())
+}
+
+#[cfg(feature = "alloc")]
+pub fn error_string(errno: __wasi_errno_t) -> String {
+    use alloc::{format, string::ToString};
+    let code = if let Some(code) = NonZeroU16::new(errno) {
+        code
+    } else {
+        return "Success".to_string();
+    };
+    match code {
+        E2BIG => "Argument list too long",
+        EACCES => "Permission denied",
+        EADDRINUSE => "Address in use",
```

Comment:
BTW do all these errors can originate from "syscalls" or are there errors which are inherited from `libc`?

Reply:
I don't have a complete list, but `EDOM` is one error code inherited from `libc` and never generated from "syscalls".

Reply:
It's possible `EOWNERDEAD` will end up in that category too, though we won't know for sure until we implement pthreads.

`EIDRM` is an error code inherited from POSIX that isn't generated by WASI, at least for the foreseeable future. Same for `ESRCH` if WASI stays on the path of avoiding the "process" concept at the API level.

Reply:
It will be worth to remove such errors later. Also it would be nice to add to the Core API [description](https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-api.md) list of possible errors to each "syscall" (ofc with a disclaimer that those lists are not complete and implementations may return other error codes).

Reply:
We also could change return type to `Option<&'static str>` to avoid allocations.